pkg/manager/member/pd_upgrader.go

```diff
@@ -118,24 +118,35 @@ func (pu *pdUpgrader) gracefulUpgrade(tc *v1alpha1.TidbCluster, oldSet *apps.Sta
 func (pu *pdUpgrader) upgradePDPod(tc *v1alpha1.TidbCluster, ordinal int32, newSet *apps.StatefulSet) error {
 	ns := tc.GetNamespace()
 	tcName := tc.GetName()
-	upgradePodName := PdPodName(tcName, ordinal)
-	if tc.Status.PD.Leader.Name == upgradePodName && tc.PDStsActualReplicas() > 1 {
-		lastOrdinal := tc.PDStsActualReplicas() - 1
+	upgradePdName := PdName(tcName, ordinal, tc.Namespace, tc.Spec.ClusterDomain)
+	if tc.Status.PD.Leader.Name == upgradePdName && len(tc.Status.PD.PeerMembers) > 1 {
 		var targetName string
-		if ordinal == lastOrdinal {
-			targetName = PdPodName(tcName, 0)
+		if tc.PDStsActualReplicas() > 1 {
+			targetOrdinal := tc.PDStsActualReplicas() - 1
+			if ordinal == targetOrdinal {
+				targetOrdinal = 0
+			}
+			if _, exist := tc.Status.PD.PeerMembers[targetName]; !exist {
+				targetName = PdName(tcName, targetOrdinal, tc.Namespace, tc.Spec.ClusterDomain)
+			} else {
+				targetName = PdName(tcName, targetOrdinal, tc.Namespace, tc.Spec.ClusterDomain)
+			}
 		} else {
-			targetName = PdPodName(tcName, lastOrdinal)
+			for _, member := range tc.Status.PD.PeerMembers {
+				if member.Name != upgradePdName && member.Health {
+					targetName = member.Name
+					break
+				}
+			}
 		}
 		err := pu.transferPDLeaderTo(tc, targetName)
```

Comment:
There is leader transfer logic in `pd_scaler.go` too and we need to update that.

Reply:
add transfer leader to peer member logic